transactions/multi-document-acid-transactions-mongodb-go.md

```diff
@@ -0,0 +1,162 @@
+
+```go
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// Episode represents the schema for the "Episodes" collection
+type Episode struct {
+	ID          primitive.ObjectID `bson:"_id,omitempty"`
+	Podcast     primitive.ObjectID `bson:"podcast,omitempty"`
+	Title       string             `bson:"title,omitempty"`
+	Description string             `bson:"description,omitempty"`
+	Duration    int32              `bson:"duration,omitempty"`
+}
+
+func main() {
+	client, err := mongo.Connect(context.TODO(), options.Client().ApplyURI(os.Getenv("ATLAS_URI")))
+	if err != nil {
+		panic(err)
+	}
+	defer client.Disconnect(context.TODO())
+
+	database := client.Database("quickstart")
+	episodesCollection := database.Collection("episodes")
+}
+```
+
+Now let's also assume that you've correctly included the MongoDB Go driver as seen in a previous tutorial titled, [How to Get Connected to Your MongoDB Cluster with Go](https://www.mongodb.com/blog/post/quick-start-golang--mongodb--starting-and-setup).
+
+The goal here will be to try to insert a document that complies with our schema validation as well as a document that doesn't, so that we have a commit that doesn't happen.
+
+```go
+// ...
+
+func main() {
+    // ...
+
+	session, err := client.StartSession()
+	if err != nil {
+		panic(err)
+	}
+	if err = session.StartTransaction(); err != nil {
+		panic(err)
+	}
+	err = mongo.WithSession(context.Background(), session, func(sessionContext mongo.SessionContext) error {
+		result, err := episodesCollection.InsertOne(
+			sessionContext,
+			Episode{
+				Title:    "A Transaction Episode for the Ages",
+				Duration: 15,
+			},
+        )
+        fmt.Println(result.InsertedID)
+		result, err = episodesCollection.InsertOne(
+			sessionContext,
+			Episode{
+				Title:    "Transactions for All",
+				Duration: 1,
+			},
+		)
+		if err = session.CommitTransaction(sessionContext); err != nil {
+			panic(err)
+        }
+        fmt.Println(result.InsertedID)
+		return nil
+	})
+	if err != nil {
+		panic(err)
+	}
+	session.EndSession(context.Background())
+}
+```
+
+In the above code we start by starting a session which will encapsulate everything we want to do with atomicity. After, we start a transaction which we'll use to commit everything in the session.
+
+Inside the session we are doing two `InsertOne` operations. The first would succeed because it doesn't violate any of our schema validation rules. It will even print out an object id when it's done. However, the second operation will fail because it is less than two minutes. Because the `CommitTransaction` won't succeed, both `InsertOne` will not exist in the database even though the first would have succeeded.
```

Comment:
I would explain how the transaction actually gets aborted because you don't call `AbortTransaction`. Does the server automatically abort it when the insert fails? I know the `EndSession` call also aborts an in-progress transaction related to that session, so maybe that's where it's happening?

Reply:
That's a good question, and one that I don't know the answer to.

My assumption is that if `CommitTransaction` never happens, the transaction is aborted, but maybe aborting is not equal to not committing.

This is where my understanding of using transactions vs knowing how they work comes into question.

Any thoughts?

Reply:
Looking at the code example in https://docs.mongodb.com/v4.0/core/transactions/#transactions-api, it seems like the recommendation is to explicitly abort the transaction if an error occurs, so I'd say we should follow that.